### ISSUE TITLE
False negative around unioned typed dictionaries that inherit from a common base class

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#9131]
I ran into a false positive situation around typed dictionaries that inherit from a common base class.

```python
from typing import TypedDict


class _Base(TypedDict): ...


class _Result1(_Base): ...


class _Result2(_Base): ...


class Container:
    def _build(self, klass: type[_Result1] | type[_Result2]) -> _Result1 | _Result2: ...

    def build(self) -> _Result1:
        # No warning, false negative.
        return self._build(_Result1)
```

Removing type dict leads to a warning being properly raised:

```python
class _Base: ...


class _Result1(_Base): ...


class _Result2(_Base): ...


class Container:
    def _build(self, klass: type[_Result1] | type[_Result2]) -> _Result1 | _Result2: ...

    def build(self) -> _Result1:
        # Warning raised
        return self._build(_Result1)
```

[Pyright Playground
](https://pyright-play.net/?pyrightVersion=1.1.383&pythonVersion=3.14&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFD1UA2AhgM6tQD6AQm8QBREEZSjQCUALigA6WfVpM2HTgCVirAK6MYARn48%2BkmXIYt2XNZu0Amfb1bEjs6fMXmAwmBQxmqYiAlaKGCoUmJgLgAjDSRGUn4HRmAAGigAazNWKXhhAG1VdS1dAF0oAB9YEnzLIutisSgAWgA%2BC0LtHXK2qxhrKWd6ENDwqGjY%2BMTgBpbuop1AoaGQYhgNEBQoSelOMbj9Go6xWiA)
Pyright Version: 1.1.383

**Comment by erictraut** [comment]
Pyright's behavior is correct here. This isn't a false negative.

A `TypedDict` is a structural type. The types `_Result` and `_Result2` are equivalent in your first code sample. If you omit `TypedDict`, these classes are nominal classes, and the two classes are not equivalent.

**Comment by max-muoto** [comment]
> Pyright's behavior is correct here. This isn't a false negative.
> 
> A `TypedDict` is a structural type. The types `_Result` and `_Result2` are equivalent in your first code sample. If you omit `TypedDict`, these classes are nominal classes, and the two classes are not equivalent.

Apologies, I should have shown an example where the subclasses were differing in structure. This is more akin to what I actually saw, and it still fails to raise a warning.

```python
from typing import TypedDict


class _Base(TypedDict): ...


class _Result1(_Base): ...


class _Result2(_Base):
   x: int


class Container:
    def _build(self, klass: type[_Result1] | type[_Result2]) -> _Result1 | _Result2: ...

    def build(self) -> _Result1:
        # No warning, false negative.
        return self._build(_Result1)
```

**Comment by erictraut** [comment]
The `_Result1` class is a TypedDict that has zero items. All other TypedDicts are compatible with this structural type. That means `_Result1 | _Result2` is assignable to `_Result1`. There is no type violation here.
